### PR TITLE
Eclipse autocompletion support

### LINKED ISSUE #5667: Improve support for Eclipse
Hi,

I made some modifications to the AppModule for the Eclipse IDE.

It's to allow using the autocompletion feature without leaving the focus from the code editor.

[eclipse_appModule_patch.txt](https://github.com/nvaccess/nvda/files/85156/eclipse_appModule_patch.txt)

It's a patch file.

Cheers.

### PR DESCRIPTION
**Pull request by francipvb**
### Link to issue number:

Closes #5667

### Summary of the issue:

The issue is that when writing code in the editor you need to focus the popup to know what items are on it. This aims to solve this particular problem.

### Description of how this pull request fixes the issue:

It detects the popup and read the selected item as it changes.

This also includes a little script to read javadoc right from the editor.

I tried to avoid relations with the OS controls.

### Testing performed:

I tested it against java and XML editors. It works well, but in XML it will not read any documentation.

### Known issues with pull request:

* Probably will not read documentation for other languages than java.

### Change log entry:

Section: New features

* Eclipse: Added support for autocompletion in code editor. #5667
  - Additionally, _Javadoc_ information can be read with *NVDA+d* right from the editor when it is present.

### CODE REVIEW

File: source/appModules/eclipse.py [hunk 1]
```diff
@@ -17,7 +26,102 @@ def event_valueChange(self):
...
+		description = _("Tries to read documentation for the selected autocompletion item."),
+		gesture = "kb:nvda+d"
+	)
+	def script_readDocumentation(self, gesture):
+		# If there aren't any suggestion selected, there is no way to find quick documentation
+		if not self.appModule.selectedItem:
+			gesture.send()
+			return
+
+		obj = self.appModule.selectedItem.parent.parent.parent.parent.previous.previous
```
Comment (feerrenrut) [inline source/appModules/eclipse.py]: This seems error prone, what can you do to make this less fragile? Is previous always going to exist? How can we verify that we have the object that we expect?

File: source/appModules/eclipse.py [hunk 2]
```diff
@@ -17,7 +26,102 @@ def event_valueChange(self):
...
+		# When writing in the editor, the selection event is not fire.
+		# This is probably because the items are reused.
+		if self.appModule.selectedItemName != self.name:
+			self.appModule.selectedItemName = self.name
+			speech.cancelSpeech()
+
+			# Reporting as focused should be sufficient
+			self.reportFocus()
+
+			# Fixme: I picked up this from UIA SuggestionItem
```
Comment (feerrenrut) [inline source/appModules/eclipse.py]: Please either remove this, or explain the problems that are caused by this code, and why it should be used for now.

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/appModules/eclipse.py (hunk 1 above)
Comment (francipvb) [reply source/appModules/eclipse.py]: Reworked this completely.

Thanks for your help.

**Comment by francipvb** [comment]
CC @leonardder 

What do you think about trying this approach for Visual Studio?

File: source/appModules/eclipse.py (hunk 2 above)
Comment (francipvb) [reply source/appModules/eclipse.py]: Addressed this, to.

Thanks again.

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/appModules/eclipse.py [hunk 3]
```diff
@@ -101,17 +131,22 @@ def event_selection(self):
...
 		if self.appModule.selectedItemName != self.name:
 			self.appModule.selectedItemName = self.name
 			speech.cancelSpeech()
 
 			# Reporting as focused should be sufficient
 			self.reportFocus()
 
-			# Fixme: I picked up this from UIA SuggestionItem
-			braille.handler.message(braille.getBrailleTextForProperties(name=self.name, role=self.role, position=self.positionInfo))
+			# Fixme: I picked up this from UIA SuggestionItem for rendering in braille.
```
[the remark this answers is not in the document]
Comment (francipvb) [reply source/appModules/eclipse.py]: Hello @feerrenrut,

No, I don't think that this is a problem. I just am unsure if this is the right way. If you consider this is the best and correct approach I can remove this comment.

I don't have a Braille display to test this.

Cheers,

File: source/appModules/eclipse.py [hunk 4]
```diff
@@ -42,34 +46,60 @@ def event_caret(self):
...
+				pass
 
-		if obj:
+		# Check if this object is from the same appModule
+		if obj and obj.appModule == self.appModule:
 			api.setNavigatorObject(obj)
-			braille.handler.handleReviewMove()
-			sayAllHandler.readText(sayAllHandler.CURSOR_REVIEW)
+
+			while obj:
```
[the remark this answers is not in the document]
Comment (francipvb) [reply source/appModules/eclipse.py]: This is the only place where this code is used, so I don't know why split this into another function.

[the remark this answers is not in the document]
Comment (feerrenrut) [reply source/appModules/eclipse.py]: Re-use is one reason to split code up, another is to improve the readability. You can hide the details, and replace it with an expressive name. This reduces the number of variables a reader has to keep track of, and provides them with a higher level at which to read the code.  This reduces the control flow in the function so the high level approach is more obvious.

**Review by feerrenrut — APPROVED** [review]

**Pull request merged**